src/java/org/apache/cassandra/transport/Dispatcher.java

```diff
@@ -77,25 +77,31 @@ void processRequest(Channel channel, Message.Request request, FlushItemConverter
         assert request.connection() instanceof ServerConnection;
         ServerConnection connection = (ServerConnection) request.connection();
         Thread requestThread = Thread.currentThread();
-        this.processRequest(connection, request, backpressure)
-            .whenCompleteAsync((response, ex) -> {
-                FlushItem<?> toFlush;
-                if (ex == null)
-                {
-                    toFlush = forFlusher.toFlushItem(channel, request, response);
-                    Message.logger.trace("Responding: {}, v={}", response, connection.getVersion());
-                }
-                else
-                {
-                    JVMStabilityInspector.inspectThrowable(ex);
-                    ExceptionHandlers.UnexpectedChannelExceptionHandler handler = new ExceptionHandlers.UnexpectedChannelExceptionHandler(channel, true);
-                    ErrorMessage error = ErrorMessage.fromException(ex, handler);
-                    error.setStreamId(request.getStreamId());
-                    toFlush = forFlusher.toFlushItem(channel, request, error);
-                }
-                ClientWarn.instance.resetWarnings();
-                flush(toFlush);
-            }, Thread.currentThread() == requestThread ? Stage.IMMEDIATE.executor() : Stage.NATIVE_TRANSPORT_REQUESTS.executor());
+        try
+        {
+            this.processRequest(connection, request, backpressure)
+                .whenCompleteAsync((response, ex) -> {
+                    FlushItem<?> toFlush;
+                    if (ex == null)
+                    {
+                        toFlush = forFlusher.toFlushItem(channel, request, response);
+                        Message.logger.trace("Responding: {}, v={}", response, connection.getVersion());
+                    }
+                    else
+                    {
+                        JVMStabilityInspector.inspectThrowable(ex);
+                        ExceptionHandlers.UnexpectedChannelExceptionHandler handler = new ExceptionHandlers.UnexpectedChannelExceptionHandler(channel, true);
+                        ErrorMessage error = ErrorMessage.fromException(ex, handler);
+                        error.setStreamId(request.getStreamId());
+                        toFlush = forFlusher.toFlushItem(channel, request, error);
+                    }
+                    flush(toFlush);
+                }, Thread.currentThread() == requestThread ? Stage.IMMEDIATE.executor() : Stage.NATIVE_TRANSPORT_REQUESTS.executor());
+        }
+        finally
+        {
+            ClientWarn.instance.resetWarnings();
```

Comment:
If this works, it's only accidental, isn't it?

In the async case we'll be resetting warnings in a random moment, possibly during the execution.
Perhaps resetting the warning before handling a request is a better solution?

Reply:
OK, so let's go through my understanding how this works now:

`ClientWarn.instance.resetWarnings();` will remove the thread-local state (`FastThreadLocal<State>` that keeps the warnings). 

So in `finally` we remove the warnings kept for the thread that runs `processRequest`. 

Switching the executor we'll copy that thread client warn state to thread locals of a different thread.

So, when we schedule the response on `Stage.NATIVE_TRANSPORT_REQUESTS.executor()` then the thread local client warn state of the relevant thread used for execution will have the client warn set filled with warnings, copied over from the earlier stages.

Then, this thread may be re-used for handling a new client request. This new client request should have an empty client warn state.
But this thread local state was never cleared after sending the response.

Am I correct?

Could we clean up the client warn state at the start of request handling? This removes all the headaches related to when to clean up etc.

Reply:
> But this thread local state was never cleared after sending the response.

I guess with this you mean we should clear the warnings on _both_ sides, and yes I should have not removed the "response" side as that's exactly the reason why I put it there in the first place. I'll fix it.

> Could we clean up the client warn state at the start of request handling? 

No, because the threads are shared across stages :/

Reply:
> No, because the threads are shared across stages :/
That should not be a problem.

We start handling a user request in NTR thread. We *know* the user warnings should be clean at this point, regardless of what was this thread used for earlier (the warnings should have been copied already by the executor magic), or what it will be used for later (again - the executor magic will overwrite the thread local warnings)

So, why isn't it perfectly safe to reset the warnings when we begin user request handling in a NTR thread?

Reply:
@jakubzytka At the moment I opted to simplify by always sticking to the same stage for the response too (the previous version didn't work anyway because the thread was captured early, silly overlooking), and cleaned up everything twice. Will come back to your proposal later.